### ISSUE TITLE
Initialising spdlog inside DLL rather than main function

### ISSUE THREAD

**Issue by IIFE** [gabime/spdlog#1530]
Hi,

If I want to use spdlog as a static lib inside a DLL, can I only initialise spdlog inside the DLL and not have to do that from the main function of the app that consumes the DLL? For example can I only initialise spdlog inside DllMain?

**Comment by tt4g** [comment]
I think you shouldn't do spdlog initialization in with `DllMain` as it has many limitations.
Many functions such as thread creation and memory allocation in DllMain do not work properly.

It is recommended to prepare an initialization function other than `DllMain`.

See Microsoft document:

* [DllMain entry point \(Process\.h\) \- Win32 apps \| Microsoft Docs](https://docs.microsoft.com/ja-jp/windows/win32/dlls/dllmain)
* [Dynamic\-Link Library Best Practices \- Win32 apps \| Microsoft Docs](https://docs.microsoft.com/ja-jp/windows/win32/dlls/dynamic-link-library-best-practices)

**Comment by IIFE** [comment]
@tt4g Thanks. Is it then possible to only use and initialise spdlog from within a DLL, without having do expose spdlog to the consuming executable?

**Comment by tt4g** [comment]
There is no problem in using it in the DLL.
If you need initialization or cleanup, add the exported initialization and cleanup functions to DLL.

**Comment by IIFE** [comment]
OK thanks. I think I now have an idea of initialising spdlog once inside a DLL and using the logger from within the DLL, without the executable knowing anything about spdlog. Here is the approach I'm thinking of (untested code):

```
class SpdLogWrapper
{
public:
   static std::shared_ptr<spdlog::logger> GetLogger()
   {
      static SpdLogWrapper init;
      return init.GetSpdLogger();
   }
private:
   SpdLogWrapper(); // In the CPP implementation initialise spdlog
   std::shared_ptr<spdlog::logger> GetSpdLogger(); // Get initialised instance of spd logger

private:
   std::shared_ptr<spdlog::logger> m_logger;
};
```

Then I would call `GetLogger` whenever I want to log something out, which should ensure the logger is only initialised once.

Might go with this approach, unless you can think of any limitations from spdlog point of view.

Thanks.

**Comment by gabime** [comment]
Seems ok. 
Note that it would be faster to call this once or twice and save the pointer in main instead of asking for it each time you need to log something.

**Comment by IIFE** [comment]
@gabime `m_logger` will be set only once inside the wrapper, which is a singleton, so the logger will be created only once. Can make `GetLogger` return a reference to the shared_ptr, which will save reference count increments.

Just wanted to ensure I can use spdlog only from a DLL and it doesn't have to be initialised in main, which seems to be the case.

Thanks for the prompt responses :)

I do have another question, which is unrelated. I'll open another issue.

Thanks again.

**Comment by gabime** [comment]
Well in that case just return a pointer.  I don't think there is any point returning ref to shared ptr.

**Comment by IIFE** [comment]
That can work too :)

**Comment by shash29-dev** [comment]
Hi, IIFE 
Can you provide an elaborative response to this. Or point me towards some good references?
